### PR TITLE
OpenCL support for cracking PGP SDAs

### PR DESCRIPTION
**Pull request by kholia**

### CODE REVIEW

File: src/opencl/pgpsda_kernel.cl [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	SHA1_Init(&ctx);
+	SHA1_Update(&ctx, salt, 8);
+	for (i = 0, j = 0; i < iterations; i++, j++) {
+		SHA1_Update(&ctx, password, plen);
+#if __ENDIAN_LITTLE__
+		SHA1_Update(&ctx, (uchar*)&j, 1);
+#else
+		SHA1_Update(&ctx, ((uchar*)&j) + 3, 1);
+#endif
+	}
```
Comment (magnumripper) [inline src/opencl/pgpsda_kernel.cl]: `i` and `j` are the same at all times. Just drop `j` and use `i`. The same goes for CPU code (and pgpdisk_fmt_plug.c also)

Comment (kholia) [reply src/opencl/pgpsda_kernel.cl]: I noticed this earlier today.

Here is the original "upstream" code.

```
PGPUInt8    j;
PGPUInt16  i, k;

for (i=0, j=0; i<k; i++, j++)
{
    pgpSHAUpdate(&sha,gpp->pszPassPhrase,strlen(gpp->pszPassPhrase));   
    pgpSHAUpdate(&sha,&j,1);
}
```

I was wrong when I said that `i` and `j` are unsigned 16-bit numbers. After double-checking the original code, `j` is an unsigned char. It seems that integer wrapping would come info affect with the original code. I wondering how our code (which now uses 32-bit numbers) is working well too.

Comment (kholia) [reply src/opencl/pgpsda_kernel.cl]: Hmm. I think that in our code (which now uses 32-bit numbers), the lowest byte is affected in the same way as the original `unsigned char j` is affected by the increment operations. I am still a bit uneasy about all this :-)

Comment (kholia) [reply src/opencl/pgpsda_kernel.cl]: This original code has a comment saying `Hash the passphrase and a rotating byte counter the specified number of times`.

Comment (kholia) [reply src/opencl/pgpsda_kernel.cl]: To convince myself that things are OK in our code, I wrote the following code,

``` C
#include <stdio.h>
#include <stdint.h>

int main()
{
	uint8_t j; // original j
	uint16_t iterations = 16000;
	uint16_t i;

	uint32_t oj; // our 32-bit j

	for (i = 0, j = 0, oj = 0; i < iterations; i++, j++, oj++) {
		uint8_t *v = (uint8_t *)&i;

		if (j != *v) {
			printf("Failed!\n");
			break;
		}
		printf("%d ", j);
	}
	printf("\n");

	return 0;
}
```

**Comment by kholia** [comment]
@magnumripper This OpenCL format works well on `Intel(R) HD Graphics Skylake ULT GT2` device on Linux. However, very similar OpenCL kernels in `pgpwde-opencl` and `pgpdisk-opencl` formats fail with `FAILED (cmp_all(1))` error message. Is there something we can do to make these failing formats work with Intel OpenCL runtime as well?

**Comment by magnumripper** [comment]
> Intel(R) HD Graphics Skylake ULT GT2 device on Linux

Please post `--list=opencl-devices` for that device!

> pgpwde-opencl and pgpdisk-opencl formats fail with FAILED (cmp_all(1)) error message. Is there something we can do to make these failing formats work with Intel OpenCL runtime as well?

I'll see if I can find Linux drivers for whatever HD Graphics device I have on that gear. On my macbook, the HD Graphics 4000 device works fine with pgpdisk and pgpsda but crashes in typical mac-OS-driver-bug way with pgpwde. I could try working around that too but that device is pretty useless anyway (much weaker than a GT650M).

**Comment by kholia** [comment]
The Intel OpenCL device is good for code development purposes.

```
$ ../run/john --device=0 --list=opencl-devices
Platform #0 name: Intel Gen OCL Driver, version: OpenCL 2.0 beignet 1.3
    Device #0 (0) name:     Intel(R) HD Graphics Skylake ULT GT2
    Device vendor:          Intel
    Device type:            GPU (LE)
    Device version:         OpenCL 2.0 beignet 1.3
    Driver version:         1.3 
    Native vector widths:   char 8, short 8, int 4, long 2
    Preferred vector width: char 16, short 8, int 4, long 2
    Global Memory:          4.0 GB
    Global Memory Cache:    8.0 KB
    Local Memory:           64.0 KB (Global)
    Max memory alloc. size: 3.0 GB
    Max clock (MHz):        1000
    Profiling timer res.:   80 ns
    Max Work Group Size:    512
    Parallel compute cores: 24
    Stream processors:      96  (24 x 4)
    Speed index:            96000
```

Thanks for taking a look 👍

**Pull request merged**